Implement a mostly-faithful dispatcher in python

bdhirsh opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70364 DispatchKeySet perf improvements
* #69633 free up dispatch key space (in C++)
* **#68743 Implement a mostly-faithful dispatcher in python**

Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

bdhirsh committed:
Implement a mostly-faithful dispatcher in python

[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"




[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"




[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"




[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"




[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"




[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"




[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

zou3519 commented on torch/_mock_dispatcher/dispatch_key.py:
```diff
@@ -0,0 +1,248 @@
+from enum import Enum, auto
+
+class DispatchKey(Enum):
```
Sorry it has been a while so I don't remember the context behind this (and I'm still reading this so apologizes if this gets answered somewhere in the PR) -- what's the purpose of adding the Mock Dispatcher? Is it to do parity checking between the PyTorch dispatcher and the Mock Dispatcher?

bdhirsh replied:
Yeah good question - there were two reasons I made this PR:

(1) It was easier to reason through and debug the changes in python first before porting them to C++
(2) I could imagine it being useful in the future to have a prototype of the dispatcher in python to test out changes in the future.

That said, this python version doesn't map completely to the C++ version - I left out a bunch of features that weren't useful to me, like schema parsing and boxing/unboxing. I also don't have any tests, since I used the python dispatcher mostly just to prove out the design and minimize the time I spent debugging C++.

I probably won't spend time cleaning this up and adding features / tests. So I'd be okay with not landing this PR (and if it would be useful in the future to play around with a python prototype of the dispatcher, I could always grab the changes directly from this PR). But if anyone has stronger feelings about landing / not landing it, I'd be okay either way.

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

bdhirsh committed:
Update on "Implement a mostly-faithful dispatcher in python"


Differential Revision: [D33255192](https://our.internmc.facebook.com/intern/diff/D33255192)

[ghstack-poisoned]

bdhirsh commented:
@bdhirsh has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33255192).

yiliu30 commented:
Hi @bdhirsh, I am currently studying Dispatcher's design philosophy and this PR is very helpful. Could you please restore the branch so I can test it locally?
